File: Leetcode/cpp/String to Integer (atoi).cc
class Solution {
/*
 * 正常的atoi不是有异常就返回0么？
 * 这题出的人干事？
*/
public:
    int atoi(const char *str)
    {
        int sig = 1;
        int i = 0;
        long long ans = 0;
        
        while(str[i] == ' ') {
            ++i;
        }
        
        if (str[i] == '-') {
            i++;
            sig = -1;
        } else if (str[i] == '+') {
            i++;
        }
        
        for(i; str[i]; i++) {
            if (str[i] >= '0' && str[i] <= '9') {
                ans = ans * 10 + str[i] - '0';
            } else {
                break;
            }
        }
        ans *= sig;
        
        if (ans > INT_MAX) {
            ans = INT_MAX;
        }
        
        if (ans < INT_MIN) {
            ans = INT_MIN;
        }
        return ans;
    }
};

File: LintCode/python/String to Integer(atoi).py
import re

INT_MAX = (1 << 31) - 1
INT_MIN = -(1 << 31)

class Solution:
    # @param str: a string
    # @return an integer
    def atoi(self, str):
        str = str.strip()
        r = re.search("^[+-]?\d+", str)
        if not r:
            return 0
        str = r.group()
        res = int(str)
        res = min(INT_MAX, res)
        res = max(INT_MIN, res)
        return res
